daisy/persistence/mongodb_graph_provider.py

```diff
@@ -160,6 +165,20 @@ def read_nodes(self, roi, attr_filter=None, read_attrs=None):
             read_attrs (``list`` of ``string``):
 
                 Attributes to return. Others will be ignored
+
+            join_collection (``string``):
+
+                Compute (left) join of the nodes collection and this
+                collection using the id attribute.
+                Some node attributes can be stored in another collection.
+                Assumes that all nodes have the same attributes.
```

Comment:
Potentially under defined. What do you mean by assuming that all nodes have the same attributes? I think in this case the attributes are the `read_attrs` right? It would also help to add that nodes that don't have all `read_attrs` are silently ignored. Maybe the description can be expanded with something like: "Attributes in nodes in the `join_collection` that have matching `id`s are joined to the main collection. Joined nodes that don't have all required `read_attrs` are silently dropped." Would this be accurate based on the provided tests?

Reply:
`read_attrs` is optional.
If `read_attrs` is set the output nodes can have **at most** the attributes in `read_attrs`. If there is an attribute in `read_attrs` but not in the `nodes` collection (or in the `join_collection`) that node will still be in the output, just without this attribute (unchanged behavior to before).
If `read_attrs` is not set and `join_collection` is not set we just take whatever is in each node.
However, in principle, we cannot really assume anything about the content of each node entry as in this type of database each entry could be completely different.
And that is a bit problematic if `read_attrs` is not set but `join_collection` is set. Due to the way entries are matched across collections in mongodb ("$lookup") we have to do a projection afterwards (otherwise the attributes from the `join_collection` are nested within an array/dict). And a projection can contain either negative entries (take all except these) or positive entries (take only these) but not both. So because we have to do a (positive) projection we have to list all attributes that we want in the output, but we do not know the attributes in the node entries. So what I am doing is taking a random entry and adding all the attributes in that entry to the projection. There might be extra attributes in this entry compared to other entries, that is ignored, but there also might be missing attributes in this entry and those will be discarded from the other entries, too.
I guess one (ugly) alternative could be to loop over all nodes at the end, and fix them manually, instead of the projection. 
I hope this makes sense, please ask if I should clarify anything

Reply:
the tasks are `lookup`, `unwind`, `projection`
`lookup` does a "left outer join".
nodes collection:
```
id:1
```
join_collection:
```
id: 1
_id: hash
scoreA: 1
scoreB: 2
```

The result of the lookup op is 
```
id: 1
joined: [{id: 1, _id: hash, scoreA: 1, scoreB: 2}]
```
The list in `joined` could contain several entries, if the id occurs several times. `unwind` would split those into separate entries. However if there is only one it just removes the list
```
id: 1
joined: {id: 1, _id: hash, scoreA: 1, scoreB: 2}
```
The projection flattens the nested dict, removes the duplicate `id` attribute and removes the `_id` attribute
```
id: 1
scoreA: 1
scoreB: 2
```

So the projection has two different tasks, depending on the arguments.
If `read_attrs` is set it should filter the attributes.
If `join_collection` is set it should flatten the nested dicts and clean up the result. However it still has the "side effect" of doing the filtering.

Reply:
Regardless of implementation (i.e., the use of `lookup`, `unwind`, etc...) at the high level I was thinking that the right thing to do is to _not_ do the filtering when `join_collection` is used but not `read_attrs`, right? Just grab all attributes in the node with same `id` in `join_collection`, join them with the ones available in `nodes_collection`, and return everything. Would that be correct?

For the case where both `join_collection` and `read_attrs` are used, do the same thing but set the correct projection so only the queried attrs are returned. I think you can do this over two separate queries, one for `nodes_collection` and another for `join_collection`. You can optimize it to be 1 single query later once the behavior is more defined.

Reply:
@trivoldus28 I pushed on update. It now uses python list/dict comprehension to perform the join etc manually instead of pymongo functions. Likely slower (especially for larger collections) but hopefully with fewer assumptions.

Reply:
@abred Ok looks good! Sorry for the delay. With this reference implementation you probably can now optimize and see how much more performance it gives you.

Just one quick question: this implementation returns a full `join`, and not just a `left join` right? I'm looking at this diagram: https://stackoverflow.com/questions/53949197/isnt-sql-a-left-join-b-just-a

Reply:
I don't think I will optimize it. As far as I can see I cannot get the identical behavior with the mongodb functionality (as it is not a relational database but a document-based one there are no guarantees for the content of each document), but I think it is fine.

No, I think it is sitll a `left join`, it takes the rows in `A` and attaches the attributes from `B`, but if there is a row in `B` that is not included in `A` it will not be part of the result.

Reply:
Okay, thanks for the clarification!